G-Research.OpenTelemetry.Processor.Partial/PartialActivityProcessor.cs

```diff
@@ -10,111 +10,148 @@ namespace GR.OpenTelemetry.Processor.Partial;
+        int processIntervalMilliseconds = DefaultProcessIntervalMilliseconds
+    )
     {
-#if NET
-        ArgumentNullException.ThrowIfNull(logExporter);
-#else
-        if (logExporter == null)
-        {
-            throw new ArgumentOutOfRangeException(nameof(logExporter));
-        }
-#endif
-        this.logExporter = logExporter;
-        logProcessor = new SimpleLogRecordExportProcessor(logExporter);
+        ValidateParameters(logExporter, heartbeatIntervalMilliseconds,
+            initialHeartbeatDelayMilliseconds,
+            processIntervalMilliseconds);
 
-        if (heartbeatIntervalMilliseconds < 1)
-        {
-            throw new ArgumentOutOfRangeException(nameof(heartbeatIntervalMilliseconds));
-        }
+        _logExporter = logExporter;
+        _logProcessor = new SimpleLogRecordExportProcessor(logExporter);
 
-        this.heartbeatIntervalMilliseconds = heartbeatIntervalMilliseconds;
+        _heartbeatIntervalMilliseconds = heartbeatIntervalMilliseconds;
+        _initialHeartbeatDelayMilliseconds = initialHeartbeatDelayMilliseconds;
+        _processIntervalMilliseconds = processIntervalMilliseconds;
 
-        activeActivities = new ConcurrentDictionary<ActivitySpanId, Activity>();
-        endedActivities = new ConcurrentQueue<KeyValuePair<ActivitySpanId, Activity>>();
+        _delayedHeartbeatActivities = new ConcurrentQueue<(ActivitySpanId, DateTime)>();
+        _readyHeartbeatActivities = new ConcurrentQueue<(ActivitySpanId, DateTime)>();
+        _activeActivities = new ConcurrentDictionary<ActivitySpanId, Activity>();
 
-        shutdownTrigger = new ManualResetEvent(false);
+        _shutdownTrigger = new ManualResetEvent(false);
 
-        exporterThread = new Thread(ExporterProc)
+        _processorThread = new Thread(ProcessQueues)
         {
             IsBackground = true,
-            Name = $"OpenTelemetry-{nameof(PartialActivityProcessor)}",
+            Name = $"OpenTelemetry-{nameof(PartialActivityProcessor)}"
         };
-        exporterThread.Start();
+        _processorThread.Start();
 
-        loggerFactory = new Lazy<ILoggerFactory>(CreateLoggerFactory);
-        logger = new Lazy<ILogger<PartialActivityProcessor>>(InitializeLogger);
+        _loggerFactory = new Lazy<ILoggerFactory>(CreateLoggerFactory);
+        _logger = new Lazy<ILogger<PartialActivityProcessor>>(InitializeLogger);
     }
 
-    private void ExporterProc()
+    public override void OnStart(Activity data)
     {
-        var triggers = new WaitHandle[] { shutdownTrigger };
+        _activeActivities[data.SpanId] = data;
+        _delayedHeartbeatActivities.Enqueue((data.SpanId,
+            DateTime.UtcNow.AddMilliseconds(_initialHeartbeatDelayMilliseconds)));
+    }
 
-        while (true)
+    public override void OnEnd(Activity data)
+    {
+        var isDelayedHeartbeatPending =
+            _delayedHeartbeatActivities.Any(span => span.SpanId == data.SpanId);
```

Comment:
this is O(n) - we should instead have a way to immediately check whether a span is delayed or not (e.g. by using 2 dictionaries: one for delayed activities and one for ready activites)

Reply:
why don't we use one dictionary instead of two? we should be able to check if span is delayed with one. if span is ready we should send log because log was already sent after delay expired.

Reply:
sure thing - this was just a suggestion!